nexus/types/src/deployment/network_resources.rs

```diff
@@ -0,0 +1,283 @@
+        match self {
+            OmicronZoneExternalIp::Floating(ip) => {
+                OmicronZoneExternalIpKey::Floating(ip.ip)
+            }
+            OmicronZoneExternalIp::Snat(snat) => {
+                OmicronZoneExternalIpKey::Snat(snat.snat_cfg)
+            }
+        }
+    }
+}
+
+/// An IP-based key suitable for uniquely identifying an
+/// [`OmicronZoneExternalIp`].
+///
+/// We can't use the IP itself to uniquely identify an external IP because SNAT
+/// IPs can have overlapping addresses.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum OmicronZoneExternalIpKey {
+    Floating(IpAddr),
+    Snat(SourceNatConfig),
+}
+
+/// Floating external IP allocated to an Omicron-managed zone.
+///
+/// This is a slimmer `nexus_db_model::ExternalIp` that only stores the fields
+/// necessary for blueprint planning, and requires that the zone have a single
+/// IP.
+#[derive(
+    Debug, Clone, Copy, PartialEq, Eq, JsonSchema, Serialize, Deserialize,
+)]
+pub struct OmicronZoneExternalFloatingIp {
+    pub id: ExternalIpUuid,
+    pub ip: IpAddr,
+}
+
+/// Floating external address with port allocated to an Omicron-managed zone.
+#[derive(
+    Debug, Clone, Copy, PartialEq, Eq, JsonSchema, Serialize, Deserialize,
+)]
+pub struct OmicronZoneExternalFloatingAddr {
+    pub id: ExternalIpUuid,
+    pub addr: SocketAddr,
+}
+
+impl OmicronZoneExternalFloatingAddr {
+    pub fn into_ip(self) -> OmicronZoneExternalFloatingIp {
+        OmicronZoneExternalFloatingIp { id: self.id, ip: self.addr.ip() }
+    }
+}
+
+/// SNAT (outbound) external IP allocated to an Omicron-managed zone.
+///
+/// This is a slimmer `nexus_db_model::ExternalIp` that only stores the fields
+/// necessary for blueprint planning, and requires that the zone have a single
+/// IP.
+#[derive(
+    Debug, Clone, Copy, PartialEq, Eq, JsonSchema, Serialize, Deserialize,
+)]
+pub struct OmicronZoneExternalSnatIp {
+    pub id: ExternalIpUuid,
+    pub snat_cfg: SourceNatConfig,
+}
+
+/// Network interface allocated to an Omicron-managed zone.
+///
+/// This is a slimmer `nexus_db_model::ServiceNetworkInterface` that only stores
+/// the fields necessary for blueprint planning.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct OmicronZoneNic {
+    pub id: VnicUuid,
+    pub mac: MacAddr,
+    pub ip: IpAddr,
+    pub slot: u8,
+    pub primary: bool,
+}
+
+/// A pair of an Omicron zone ID and an external IP.
+///
+/// Part of [`OmicronZoneNetworkResources`].
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct OmicronZoneExternalIpEntry {
+    pub zone_id: OmicronZoneUuid,
+    pub ip: OmicronZoneExternalIp,
+}
+
+/// Specification for the tri-map of Omicron zone external IPs.
+impl TriMapEntry for OmicronZoneExternalIpEntry {
+    type K1 = OmicronZoneUuid;
+    type K2 = ExternalIpUuid;
+
+    // Note: cannot use IpAddr here, because SNAT IPs can overlap as long as
+    // their port blocks are disjoint.
```

Comment:
I think this is a pretty strong argument for your suggestion that we change SNAT configs to take an enum describing the block instead of a pair of `u16`s. We should no longer be able to construct SNAT configs that don't exactly match one of the port blocks, but this code is relying on that I think? (E.g., if we could still construct invalid port blocks that overlapped but weren't equal, that wouldn't be detected by this map's dup checks.)

Reply:
That's right, yeah. Maybe I should fix that then.

Reply:
Ehh, I think for now I'll just rely on `SourceNatConfig` externally guaranteeing that the port blocks are disjoint.